### ISSUE TITLE
Storing a device name with the token

### ISSUE THREAD

**Issue by tarjei** [gesdinet/JWTRefreshTokenBundle#282]
Hi, is it possible to store a device name with a given token - for example using an extra event listener on token creation that can fetch the device name from the request?

**Comment by mbabker** [comment]
With 1.1 this isn't all that bad to pull off:

1) Make sure you've got an entity set up in your app adding the device name field
2) Replace or decorate the `gesdinet.jwtrefreshtoken.refresh_token_generator` service in your app with a custom implementation of `Gesdinet\JWTRefreshTokenBundle\Generator\RefreshTokenGeneratorInterface` which adds the device name to the generated entity class

**Comment by markitosgv** [comment]
It might be a good idea to store some kind of unique hash per device/browser/ip/location to be able to show a list to users of which devices are connected and the last renewal and even to be able to revoke specifically the tokens needed.

How do you see it @mbabker ?

**Comment by shakaran** [comment]
@tarjei, yes — @mbabker's approach is the right one and still works. One correction: the service was renamed in 2.0, so it is now `gesdinet_jwt_refresh_token.refresh_token_generator`.

`RefreshTokenGeneratorInterface` is aliased, so a decorator is autowired without naming the service at all:

```php
#[AsDecorator('gesdinet_jwt_refresh_token.refresh_token_generator')]
final class DeviceAwareRefreshTokenGenerator implements RefreshTokenGeneratorInterface
{
    public function __construct(
        private readonly RefreshTokenGeneratorInterface $decorated,
        private readonly RequestStack $requestStack,
    ) {
    }

    public function createForUserWithTtl(UserInterface $user, int $ttl): RefreshTokenInterface
    {
        $token = ($this->decorated)->createForUserWithTtl($user, $ttl);
        $token->setDeviceName($this->requestStack->getCurrentRequest()?->headers->get('User-Agent'));

        return $token;
    }
}
```

The generator has no request of its own, which is why the `RequestStack` is injected. The listener calls the generator and then saves what it returns, so anything set here is stored with the token. Your own class goes in `refresh_token_class`, and [mapping the entity yourself](https://github.com/markitosgv/JWTRefreshTokenBundle#mapping-the-entity-yourself) is now documented, which is what adding a column needs.

@markitosgv, on your question to @mbabker about a device list with selective revocation — most of it is now in place, and one piece is not:

- Extra columns on the token: yes, as above.
- Bounding how many sessions a user has: `max_tokens_per_user`, added this cycle.
- Revoking every session of a user: `RevokeRefreshTokenManagerInterface::revokeAllForUser()`.
- Revoking one session: `delete()` with that token.
- **Listing a user's sessions: not offered.** `RefreshTokenManagerInterface` has `getLastFromUsername()`, which returns one. Nothing returns all of a user's tokens, so a "your devices" screen has to query the repository directly, which applications can do since they own the entity, but the bundle does not help.

That last one is the only real gap, and it is a small interface addition on the same pattern as the others. Worth doing if there is appetite for it.

Two things worth being clear about before anyone builds this. A device name sent by the client is self-reported — useful for showing a user "Chrome on Linux, last used Tuesday" so they can spot a session they do not recognise, and useless as a security control, since an attacker sends whatever string they like. And a `User-Agent` is a poor identifier: it changes on every browser update, so the same device shows up repeatedly in the list. Both are fine for the recognise-and-revoke use case, which is what makes that feature worth having; neither is a substitute for `single_use: true`.

Closing this as answered — the recipe works, and the listing gap is noted.

**Comment by shakaran** [comment]
Follow-up: the gap I mentioned above is closed, in dc855f2.

```php
public function sessions(UserInterface $user): array
{
    return array_filter(
        $this->refreshTokens->findAllForUser($user),
        static fn (RefreshTokenInterface $token): bool => $token->isValid()
    );
}
```

`Gesdinet\JWTRefreshTokenBundle\Model\ListRefreshTokenManagerInterface` is aliased to the manager, so it injects by type on every backend. Tokens come back with the one expiring last first — the same order `max_tokens_per_user` keeps from — and expired ones are included, because they are still rows in the table and a caller counting sessions should not disagree with a caller clearing it. `isValid()` is what tells them apart, as above.

That makes @markitosgv's device list buildable end to end: a token class of your own for the columns, a decorator around the generator to fill them in, `findAllForUser()` to show them, `delete()` to revoke one, `revokeAllForUser()` to revoke all, and `max_tokens_per_user` to bound how many there can be.

It also turned up something I had left half-done: the DBAL manager had gained `RevokeRefreshTokenManagerInterface` for the token limit, but its alias was still only registered for the ORM and the ODM, so it could not be injected by type there. Both interfaces are now aliased for every backend.

All of it ships in the next release.
